node/src/components/linear_chain_sync.rs

```diff
@@ -675,9 +676,14 @@ where
                             return fatal!(effect_builder, "block height mismatch").ignore();
                         }
                         trace!(%block_height, "Linear block found in the local storage.");
-                        // When syncing descendants of a trusted hash, we might have some of
-                        // them in our local storage. If that's the case, just continue.
-                        self.block_downloaded(rng, effect_builder, &block)
+                        // We need to update internal state of the `LinearChainSync` component
+                        // but we ignore the usual effects. If that block is already in storage,
+                        // it means we must have "acted" on it and processed these effects in the
+                        // past. Now, we can continue downloading the next
+                        // block in the chain - one with height + 1.
+                        let _ = self.block_downloaded(rng, effect_builder, &block);
+                        let _ = self.block_handled(rng, effect_builder, &block);
+                        self.fetch_next_block(effect_builder, rng, &block)
```

Comment:
Just to make sure we're not missing anything:

* The "incorrect state" `fatal` effect should be unreachable anyway, right?
* Actually, we're guaranteed to be in `SyncingDescendants` state?
* Fetching deploys and executing the block are no-ops because we already have it… and we store it only after execution, even if we downloaded it, right?

But the downgrade happens only when handling a `GetDeploysResult`, not in `block_handled`, does it? So, would this alternative be equivalent, and avoid throwing away effects?

```suggestion
                        self.peers.reset(rng);
                        self.state.block_downloaded(block);
                        self.add_block(block.clone());
                        self.block_handled(rng, effect_builder, &block)
```

Reply:
Actually, you might be correct - if we can guarantee that we're in `SyncingDescendants` state AND that having block in storage means it was executed, then your snippet should be equivalent. I will double-check. Good catch, thanks.

Reply:
After discussions - we agreed to not implement this alternative as exposing more low-level state manipulations makes it easier to create a bug (where order of lines/calls matters for the final result). Using higher level methods (like `block_downloaded` and `block_handled`) hides the complexity and tries to enforce calls ordering.